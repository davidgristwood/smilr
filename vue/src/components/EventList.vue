<template>
  <div>
    <transition-group name="fade" tag="p">
      <div v-for="event in events" :key="event._id">
        <b-card :border-variant="variant" :header-bg-variant="variant" header-text-variant="white">
          <h2 slot="header">
            <fa :icon="utilsFaIcon(event.type)" /> &nbsp; {{ event.title }}
          </h2>
          {{ event.start | moment("dddd, MMMM Do YYYY") }}
          <hr>
          <ul>
            <li v-for="topic in event.topics" :key="topic.id">
              <router-link v-if="filter == 'active'" :to="{ name: 'feedback', query: { e: event._id, t: topic.id }}">
                {{ topic.desc }}
              </router-link>
              <span v-else>{{ topic.desc }}</span>
            </li>
          </ul>
        </b-card>
        <br>
      </div>
    </transition-group>

    <b-alert v-if="events && events.length <= 0" show variant="info">
      <h2>No {{ filter }} events found</h2>
    </b-alert>

    <spinner v-if="!events" />
  </div>
</template>

<script>
import '../assets/css/animate.css'
import api from '../mixins/api'
import utils from '../mixins/utils'
import Spinner from './Spinner'

export default {
  name: 'EventList',

  components: {
    Spinner
  },

  mixins: [ api, utils ],

  props: {
    filter: {
      type: String,
      required: true,
    }
  },

  data: function() {
    return {
      events: null
    }
  },

  computed: {
    variant: function () {
      return this.filter == 'active' ? 'primary' : 'secondary'
    }
  },

  watch: {
    filter: function() {
      this.events = null
      this.refresh()
    }
  },

  created: function() {
    this.refresh()
  },

  methods: {
    refresh: function() {
      this.apiGetEventsFiltered(this.filter)
        .then((resp) => {
          if (resp && resp.data) {
            this.events = resp.data
          }
        })
        .catch((err) => {
          // Never should get here
          console.log(err)
        })
    }
  }
}
</script>

<style>
.card-body {
  font-size: 1.4rem;
}

.icon {
  height: 40px;
}

.fade-enter-active {
  animation: fadeInRight 0.8s;
}

.spinner {
   font-size: 6rem;
   width: 9rem;
}
</style>

