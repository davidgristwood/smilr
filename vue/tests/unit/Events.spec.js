import { shallowMount } from '@vue/test-utils'
import Events from '@/components/Events.vue'
import EventList from '@/components/EventList.vue'
import flushPromises from 'flush-promises'

jest.mock('@/mixins/api')

describe('Events.vue', () => {
  it('shows active events', async () => {
    const wrapper = shallowMount(Events, { stubs: {'EventList': EventList} })

    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })

  it('shows past events', async () => {
    const wrapper = shallowMount(Events, { stubs: {'EventList': EventList} })
    wrapper.setData({ time: 'past' })

    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })

  it('shows future events', async () => {
    const wrapper = shallowMount(Events, { stubs: {'EventList': EventList} })
    wrapper.setData({ time: 'future' })

    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })    
})
