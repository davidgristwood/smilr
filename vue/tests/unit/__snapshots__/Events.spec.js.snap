// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events.vue shows active events 1`] = `
<div>
  <h1>Browse Events</h1>
  <b-form-group label="Find events by time range" description="Select a time range filter">
    <b-form-select size="lg" class="mb-3">
      <option value="null" disabled="disabled">
        -- Please select an option --
      </option>
      <option value="past">
        Past
      </option>
      <option value="active">
        Active
      </option>
      <option value="future">
        Future
      </option>
    </b-form-select>
  </b-form-group>
  <div>
    <transition-group-stub tag="p" name="fade">
      <div>
        <b-card border-variant="primary" header-bg-variant="primary" header-text-variant="white">
          <h2 slot="header">
            <fa icon="chalkboard-teacher"></fa> &nbsp; Fake event (current)
          </h2>
          1997-01-01
          <hr>
          <ul>
            <li>
              <router-link to="[object Object]">
                Lemons
              </router-link>
            </li>
            <li>
              <router-link to="[object Object]">
                Apples
              </router-link>
            </li>
          </ul>
        </b-card> <br>
      </div>
    </transition-group-stub>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`Events.vue shows future events 1`] = `
<div>
  <h1>Browse Events</h1>
  <b-form-group label="Find events by time range" description="Select a time range filter">
    <b-form-select size="lg" class="mb-3">
      <option value="null" disabled="disabled">
        -- Please select an option --
      </option>
      <option value="past">
        Past
      </option>
      <option value="active">
        Active
      </option>
      <option value="future">
        Future
      </option>
    </b-form-select>
  </b-form-group>
  <div>
    <transition-group-stub tag="p" name="fade">
      <div>
        <b-card border-variant="secondary" header-bg-variant="secondary" header-text-variant="white">
          <h2 slot="header">
            <fa icon="flask"></fa> &nbsp; Fake lab (future)
          </h2>
          2025-01-01
          <hr>
          <ul>
            <li><span>Jetpacks</span></li>
            <li><span>Teleporters</span></li>
          </ul>
        </b-card> <br>
      </div>
    </transition-group-stub>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`Events.vue shows past events 1`] = `
<div>
  <h1>Browse Events</h1>
  <b-form-group label="Find events by time range" description="Select a time range filter">
    <b-form-select size="lg" class="mb-3">
      <option value="null" disabled="disabled">
        -- Please select an option --
      </option>
      <option value="past">
        Past
      </option>
      <option value="active">
        Active
      </option>
      <option value="future">
        Future
      </option>
    </b-form-select>
  </b-form-group>
  <div>
    <transition-group-stub tag="p" name="fade">
      <div>
        <b-card border-variant="secondary" header-bg-variant="secondary" header-text-variant="white">
          <h2 slot="header">
            <fa icon="calendar-alt"></fa> &nbsp; Fake event (past)
          </h2>
          1996-01-01
          <hr>
          <ul>
            <li><span>Spice Girls</span></li>
            <li><span>Can Of Stella</span></li>
          </ul>
        </b-card> <br>
      </div>
    </transition-group-stub>
    <!---->
    <!---->
  </div>
</div>
`;
